Modules/posixmodule.c

```diff
@@ -13535,46 +13539,33 @@ struct constdef {
 };
 
 static int
-conv_confname(PyObject *arg, int *valuep, struct constdef *table,
-              size_t tablesize)
-{
-    if (PyLong_Check(arg)) {
-        int value = PyLong_AsInt(arg);
-        if (value == -1 && PyErr_Occurred())
-            return 0;
-        *valuep = value;
-        return 1;
-    }
-    else {
-        /* look up the value in the table using a binary search */
-        size_t lo = 0;
-        size_t mid;
-        size_t hi = tablesize;
-        int cmp;
-        const char *confname;
-        if (!PyUnicode_Check(arg)) {
-            PyErr_SetString(PyExc_TypeError,
-                "configuration names must be strings or integers");
-            return 0;
-        }
-        confname = PyUnicode_AsUTF8(arg);
-        if (confname == NULL)
-            return 0;
-        while (lo < hi) {
-            mid = (lo + hi) / 2;
-            cmp = strcmp(confname, table[mid].name);
-            if (cmp < 0)
-                hi = mid;
-            else if (cmp > 0)
-                lo = mid + 1;
-            else {
-                *valuep = table[mid].value;
-                return 1;
+conv_confname(PyObject *module, PyObject *arg, int *valuep, const char *tablename)
+{
+    if (PyUnicode_Check(arg)) {
+        PyObject *table = PyObject_GetAttrString(module, tablename);
+        if (table != NULL) {
+            arg = PyObject_GetItem(table, arg);
+            if (arg == NULL) {
+                PyErr_SetString(
+                    PyExc_ValueError, "unrecognized configuration name");
             }
+            Py_DECREF(table);
         }
-        PyErr_SetString(PyExc_ValueError, "unrecognized configuration name");
-        return 0;
+        if (PyErr_Occurred())
```

Comment:
Why do we need to check this? Seems like it should be sufficient to return 0 if `PyObject_GetAttrString`. `PyErr_Occurred()` is relatively expensive and we should avoid calling it if possible.

Reply:
OK, I've rearranged the code to avoid `PyErr_Occurred`, and it's ended up clearer than before.

Why is `PyErr_Occurred` expensive, though? It looks like it's just checking a member in the current thread state.